tests/test_morphing.py

```diff
@@ -0,0 +1,109 @@
+
+    # Test with the restriction of max_overall_power == 1
+    components_max = morpher.find_components(1)
+    expected_max = np.array([[0,0], [0,1], [1,0]])
+    assert components_max.any() == expected_max.any(), "The max_power of outputs of original_find_components are not restricted to one"
+
+
+def original_set_basis(basis_numpy = np.array([[1, -5], [1, -4], [1, -3], [1, -2], [1, -1]]), 
+                        parameter_max_power = [2,2], 
+                        this_components = np.array([[4, 0], [3, 1], [2, 2], [1, 3], [0, 4]]) ):
+
+    # test set basis with input of basis_numpy.
+    morpher = m.PhysicsMorpher(parameter_max_power=parameter_max_power)
+    morpher.set_components(this_components)
+    morpher.set_basis(basis_numpy = basis_numpy)
+    
+    # Check if the basis is the same
+    assert morpher.basis.any() == basis_numpy.any()
+
+    expected_matrix = np.array([[1, 2.08333, 1.45833, 0.416667, 0.0416667], 
+                            [-5, -10.1667, -6.83333, -1.83333, -0.166667], 
+                            [10, 19.5, 12.25, 3, 0.25], 
+                            [-10, -17.8333, -9.83333, -2.16667, -0.166667], 
+                            [5, 6.41667, 2.95833, 0.583333, 0.0416667]]).T
+
+    # the set_basis also calls calculate matrix if self.morphing_matrix is not defined, check if the morphing matrix is expected 
+    assert morpher.morphing_matrix.any() == expected_matrix.any(), "The calculated matrix in set_basis is different from expected."
+
+def original_calculate_matrix(this_basis = np.array([[1, -5], [1, -4], [1, -3], [1, -2], [1, -1]]), 
+                                this_components = np.array([[4, 0], [3, 1], [2, 2], [1, 3], [0, 4]]),):
+
+    morpher = m.PhysicsMorpher(parameter_max_power=[2,2])
+
+    # Set the powers of components and basis
+    morpher.set_components(this_components)
+    morpher.set_basis(basis_numpy= this_basis)
+
+    # Calculate the morphing matrix
+    morphing_matrix = morpher.calculate_morphing_matrix()
+    
+    # The expected output values of the matrix based on the inputes. 
+    expected_matrix = np.array([[1, 2.08333, 1.45833, 0.416667, 0.0416667], 
+                                [-5, -10.1667, -6.83333, -1.83333, -0.166667], 
+                                [10, 19.5, 12.25, 3, 0.25], 
+                                [-10, -17.8333, -9.83333, -2.16667, -0.166667], 
+                                [5, 6.41667, 2.95833, 0.583333, 0.0416667]]).T
+    
+    # assert the matrix is the same as expected
+    assert morphing_matrix.all() == expected_matrix.all(), "The output matrix is different from expected"
+
+    output_weights = morpher.calculate_morphing_weights(theta=[1,1])
+    # Test the if the output weights are as expected
+    expected_weights = np.array([5, -24, 45, -40, 15])
+    assert output_weights.all() == expected_weights.all(), "The output weights are not as expected"
```

Comment:
Similarly to the above, `.all()` checks essentially whether all values are non-zero, so this should also be `np.array_equal` (or `np.allclose` for cases where you expect floating point variations).

Reply:
Sure, will do, thank you